R/reporter-junit.r

```diff
@@ -0,0 +1,145 @@
+#' written to a file (or stdout) in jUnit XML format. This can be read by
+#' the Jenkins Continuous Integration System to report on a dashboard etc.
+#' Requires the XML package.
+#'
+#' This works for \code{\link{expect_that}} but not for the wrappers like
+#' \code{\link{expect_equal}} etc.
+#'
+#' To fit into the jUnit structure, context() becomes the \code{<testsuite>}
+#' name as well as the base of the \code{<testcase> classname}. The
+#' test_that() name becomes the rest of the \code{<testcase> classname}.
+#' The deparsed expect_that() call becomes the \code{<testcase>} name.
+#' On failure, the message goes into the \code{<failure>} node message
+#' argument (first line only) and into its text content (full message).
+#'
+#' Execution time and some other details are also recorded.
+#'
+#' References for the jUnit XML format:
+#' http://ant.1045680.n5.nabble.com/schema-for-junit-xml-output-td1375274.html
+#' http://windyroad.org/dl/Open%20Source/JUnit.xsd
+#'
+#' Output drawn from the SummaryReporter is printed to the standard
+#' error stream during execution.
+#'
+#' @export
+#' @export JUnitReporter
+#' @exportClass JUnitReporter
+#' @aliases JUnitReporter-class
+#' @keywords debugging
+JUnitReporter <- setRefClass("JUnitReporter", contains = "Reporter",
+  fields = list(
+    "file" = "character",
+    "results" = "list",
+    "timer" = "ANY"),
+
+  methods = list(
+    initialize = function(file = "", ...) {
+      if (!require("XML", quietly = TRUE)) {
+        stop("Please install the XML package", call. = FALSE)
+      }
+      callSuper(...)
+      file <<- file
+      results <<- list()
+    },
+
+    start_reporter = function() {
+      callSuper()
+      results <<- list()
+      timer <<- proc.time()
+      context <<- "(ungrouped)"
+    },
+
+    start_context = function(desc) {
+      callSuper(desc)
+      cat(desc, ": ", file = stderr())
+    },
+
+    end_context = function() {
+      callSuper()
+      cat("\n", file = stderr())
+    },
+
+    add_result = function(result) {
+      if (result$passed) {
+        cat(colourise(".", as = "passed"), file = stderr())
+      } else {
+        failed <<- TRUE
+        if (result$error) {
+          cat(colourise("F", as = "error"), file = stderr())
+        } else {
+          cat(colourise("E", as = "error"), file = stderr())
+        }
+      }
+      result$time <- round((proc.time() - timer)[["elapsed"]], 2)
+      timer <<- proc.time()
+      result$test <- if (is.null(test) || test == "") "(unnamed)" else test
+      result$call <- if (is.null(result$call)) "(unexpected)" else result$call
+      results[[context]] <<- append(results[[context]], list(result))
+    },
+
+    end_reporter = function() {
+      cat("\n", file = stderr())
+      xmlNodeOK <- function(name, ..., attrs = NULL) {
+        ## do XML entity substitutions
+        if (!is.null(attrs)) {
+          attrs <- sapply(attrs, function(x) toString(xmlTextNode(x)))
```

Comment:
Please replace all uses of `sapply()` with the equivalent `vapply()` - that's much safer inside functions